src/workerd/io/io-own.c++

```diff
@@ -33,6 +33,38 @@ void DeleteQueue::scheduleDeletion(OwnedObject* object) const {
   }
 }
 
+void DeleteQueue::scheduleAction(jsg::Lock& js, kj::Function<void(jsg::Lock&)>&& action) const {
+  if (IoContext::hasCurrent() && IoContext::current().deleteQueue.get() == this) {
+    // Same thread! No need to enqueue.
+    action(js);
+  } else {
+    auto lock = crossThreadDeleteQueue.lockExclusive();
+    KJ_IF_SOME(state, *lock) {
+      state.actions.add(kj::mv(action));
+      KJ_REQUIRE_NONNULL(state.crossThreadFulfiller)->fulfill();
+    } else {
+      // The queue was deleted, likely because the IoContet was destroyed and the
+      // DeleteQueuePtr was invalidated. We aren't going to run the action in this
+      // case and it will be dropped on the floor. If there is an active IoContext,
+      // even if it is the wrong one, we should log a warning.
+      if (IoContext::hasCurrent()) {
+        auto err = v8::Exception::Error(
+            js.str("A promise was resolved or rejected from a different request context than "
+                   "the one it was created in. However, the creating context has already been "
+                   "destroyed and continuations for that request likely will not run correctly"_kj))
+                       .As<v8::Object>();
+        jsg::check(err->Set(js.v8Context(), js.str("name"_kj), js.str("Warning"_kj)));
+        auto stack = jsg::check(err->Get(js.v8Context(), js.str("stack"_kj)));
+
+        // Safe to mutate here since we have the exclusive lock on the queue above.
+        IoContext::current().logWarning(kj::str(stack));
+        // TODO(soon): Add documentation link to this warning.
+      }
+      action(js);
```

Comment:
Why call the action in the wrong context here? Shouldn't we simply drop it without calling it?

Reply:
Discussed in chat... summary of outcome: We'll introduce a compat flag for this entire change. With the flag off, the current behavior of running these actions in the wrong IoContext will be preserved, just in case. With the compat flag set the actions will be dropped on the floor if the original IoContext has been destroyed.